src/treemodel/selection.js

```diff
@@ -0,0 +1,261 @@
+/**
+ * @license Copyright (c) 2003-2015, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+'use strict';
+
+CKEDITOR.define( [
+	'treemodel/liverange',
+	'treemodel/attributelist',
+	'emittermixin',
+	'utils',
+	'ckeditorerror'
+], ( LiveRange, AttributeList, EmitterMixin, utils, CKEditorError ) => {
+	/**
+	 * Represents a selection that is made on nodes in {@link treeModel.Document}. Selection instance is
+	 * created by {@link treeModel.Document}. In most scenarios you should not need to create an instance of Selection.
+	 *
+	 * @class treeModel.Selection
+	 */
+	class Selection {
+		/**
+		 * Creates an empty selection.
+		 *
+		 * @constructor
+		 */
+		constructor() {
+			/**
+			 * List of attributes set on current selection.
+			 *
+			 * @private
+			 * @property {treeModel.AttributeList} _attrs
+			 */
+			this._attrs = new AttributeList();
+
+			/**
+			 * Stores all ranges that are selected.
+			 *
+			 * @private
+			 * @property {Array.<LiveRange>}
+			 */
+			this._ranges = [];
+
+			/**
+			 * Specifies whether the last added range was added as a backward or forward range.
+			 *
+			 * @private
+			 * @property {Boolean}
+			 */
+			this._lastRangeBackward = false;
```

Comment:
Side note – does it have to be "backward"? I prefer if we kept a positive properties (makes thinking easier :D).